jin/[2주차 2] GC 1편 - GC trade-off와 safepoint.md

```diff
@@ -0,0 +1,88 @@
+## 1. 자유로운 JVM을 봐. 자유로워 <br>
+
+오라클은 JVM에 대해 그리 많은 것을 정해두지 않는다. **"스펙"으로서** 존재한다. <br>
+마치 인터페이스를 선언하듯, 단지 **"OS에 맞게 바이트 코드를 실행할 수 있는" 추상적인 형태의 스택 기반 해석 머신** 이라고만 정해두었고, 다양한 회사들이 그 "구현체"를 만들어 두었다. <br> <br>
```

Comment:
추상적인 형태의 스택 기반 해석 머신.. 추상적인 형태의 스택 기반 해석 머신.. 메모..

Reply:
ㅋㅋㅋㅋㅋㅋ 급하게 적느라 모호한 표현을 그대로 방치했네.. 

나는 옛날에 공부할 때,
내부적으로 바이트 코드 번역하고 연산하는 과정에서 Operand Stack 사용해서 "스택 기반 머신"이라고 부르지 않나 추측중인데 이거 맞나?? 

JVM 마스터 홍석 도와줘~~
(Native Method Stack도 상관 있나?)

Reply:
삐삑 안녕하세요 저는 챗 쥐피티입니다. 아래는 검색 결과입니다.

## JVM이 stack-based machine인가?
Yes, the Java Virtual Machine (JVM) is a stack-based machine. Unlike register-based machines, where the instructions are executed with operands primarily located in the registers of the CPU, a stack-based machine like the JVM performs operations using a last-in-first-out (LIFO) stack.

In the context of the JVM, each thread has its own stack, known as the JVM stack, created at the same time as the thread. The stack stores frames, where each frame represents a method call and contains data such as local variables, operand stack, and reference to the runtime constant pool of the class of the current method.

- **Operand Stack**: This is where intermediate operation results are stored and manipulated. For example, when executing an arithmetic addition, the JVM pushes the operands onto the operand stack, performs the addition, and then pushes the result back onto the stack.

- **Local Variables Array**: Each frame contains an array of local variables accessed via index. These variables can hold values of types such as int, float, reference types, etc., and are used during the execution of the method.

- **Execution Engine**: When executing instructions, the JVM uses the operand stack for operations. For instance, to add two integers, it pops the two top integers from the stack, adds them, and pushes the result back onto the stack.

The stack-based nature of the JVM simplifies its design and makes it easier to implement across different architectures. This is because the JVM abstracts the details of the underlying hardware, providing a uniform execution environment for Java applications regardless of the hardware they run on.

정리하면 JVM context에서는 각각의 thread가 JVM stack을 가지고 있고, 스레드는 operand stack, 로컬 변수, 메소드 reference를 포함하는 프레임을 저장하고 LIFO 방식으로 꺼내 사용되기 때문에 stack-based machine이라고 한다고 하네여~~?? 

진호 덕분에 공부 너무 많이 된다~ 최고최고~~

Reply:
이야 맛있다~ 냠냠굿~~